File: GuideObjC/GuideObjC/ViewControllers/FeedController/FeedController.m
//
//  FeedController.m
//  GuideObjC
//
//  Created by LuzanovRoman on 12.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "FeedController.h"
#import "UIViewController+Additions.h"
#import "EGFHumanName+Additions.h"
#import "ProgressController.h"
#import "SimpleFileManager.h"
#import "UIColor+Additions.h"
#import "FeedProgressCell.h"
#import "SearchDownloader.h"
#import "EdgeDownloader.h"
#import "PostController.h"
#import "EGF2.h"

@interface FeedController ()
@property (retain, nonatomic) IBOutlet UIBarButtonItem *searchButton;
@property (retain, nonatomic) IBOutlet UIBarButtonItem *createPostButton;
@property (retain, nonatomic) IBOutlet UISegmentedControl *listSegment;
@property (retain, nonatomic) BaseDownloader *activeDownloader;
@property (retain, nonatomic) SearchDownloader *searching;
@property (retain, nonatomic) EdgeDownloader *timeline;
@property (retain, nonatomic) EdgeDownloader *feed;
@property (retain, nonatomic) NSMutableDictionary * cellHeights;
@property (retain, nonatomic) UISearchBar * searchBar;
@property (retain, nonatomic) NSString *currentUserId;
@property (retain, nonatomic) NSArray *expand;
@property (retain, nonatomic) NSString *edge;
@end

@implementation FeedController

- (void)viewDidLoad {
    [super viewDidLoad];
    _cellHeights = [NSMutableDictionary dictionary];
    _expand = @[@"creator",@"image"];
    _edge = @"posts";
    
    _searchBar = [[UISearchBar alloc] init];
    _searchBar.delegate = self;
    _searchBar.showsCancelButton = true;
    _searchBar.tintColor = [UIColor hexColor:0x5E66B1];
    
    EGF2SearchParameters * parameters = [EGF2SearchParameters parametersWithObject:@"post"];
    parameters.fields = @[@"desc"];
    parameters.expand = _expand;
    _searching = [[SearchDownloader alloc] initWithParameters:parameters];
    
    [self.graph userObjectWithCompletion:^(NSObject * object, NSError * error) {
        if ([object isKindOfClass:[EGFUser class]]) {
            _currentUserId = [object valueForKey:@"id"];
            _timeline = [[EdgeDownloader alloc] initWithSource:_currentUserId edge:@"timeline" expand:_expand];
            _feed = [[EdgeDownloader alloc] initWithSource:_currentUserId edge:_edge expand:_expand];
            _feed.tableView = self.tableView;
            [_feed getNextPage];
            _activeDownloader = _feed;
        }
    }];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if (_searchBar.text.length > 0) {
        [_searchBar becomeFirstResponder];
    }
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.destinationViewController isMemberOfClass:[PostController class]] && [sender isKindOfClass:[UITableViewCell class]]) {
        PostController * postController = segue.destinationViewController;
        NSIndexPath * indexPath = [self.tableView indexPathForCell:sender];
        postController.post = (EGFPost *)[_activeDownloader objectAtIndex:indexPath.row];
        [_searchBar resignFirstResponder];
    }
}

- (void)setActiveDownloader:(EdgeDownloader *)activeDownloader {
    if (_activeDownloader) {
        _activeDownloader.tableView = nil;
    }
    _activeDownloader = activeDownloader;
    
    if (_activeDownloader) {
        _activeDownloader.tableView = self.tableView;
    }
}

- (IBAction)changeList:(UISegmentedControl *)sender {
    if (sender.selectedSegmentIndex == 0) {
        self.activeDownloader = _feed;
    }
    else if (sender.selectedSegmentIndex == 1) {
        self.activeDownloader = _timeline;
    }
}

// MARK:- Searching
- (IBAction)beginSearch:(UISegmentedControl *)sender {
    self.activeDownloader = _searching;
    self.navigationItem.rightBarButtonItem = nil;
    self.navigationItem.leftBarButtonItem = nil;
    self.navigationItem.titleView = _searchBar;
    [_searchBar becomeFirstResponder];
}

- (void)endSearch {
    [self changeList:_listSegment];
    _searchBar.text = nil;
    self.navigationItem.rightBarButtonItem = _createPostButton;
    self.navigationItem.leftBarButtonItem = _searchButton;
    self.navigationItem.titleView = _listSegment;
}

// MARK:- UISearchBarDelegate
- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar {
    [self endSearch];
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar {
    if (searchBar.text.length == 0) {
        [self endSearch];
    }
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    [_searching showObjectsWithQuery:searchText];
}

// MARK:- FeedPostCellDelegate
- (NSString *)authorizedUserId {
    return _currentUserId;
}

- (void)deletePost:(EGFPost *)post {
    if (_currentUserId && post.id) {
        [self showConfirmWithTitle:@"Warning" message:@"Really delete?" action:^{
            [ProgressController show];
            [self.graph deleteObjectWithId:post.id forSource:_currentUserId fromEdge:@"posts" completion:^(id object, NSError * error) {
                [ProgressController hide];
            }];
        }];
    }
}

// MARK:- UITableViewDelegate
- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView {
    if (self.refreshControl.isRefreshing) {
        [_activeDownloader refreshList];
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 0) {
        EGFPost * post = (EGFPost *)[_activeDownloader objectAtIndex:indexPath.row];
        // Check if we already have the value
        NSNumber * height = _cellHeights[post.id];
        
        if (height) {
            return [height floatValue];
        }
        CGFloat newHeight = [FeedPostCell heightForPost:post];
        _cellHeights[post.id] = [NSNumber numberWithFloat:newHeight];
        return newHeight;
    }
    return 44;
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 1 && ![_activeDownloader isDownloaded]) {
        [_activeDownloader getNextPage];
    }
}

// MARK:- UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return section == 0 ? _activeDownloader.count : 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.section == 1) {
        FeedProgressCell * cell = [tableView dequeueReusableCellWithIdentifier:@"ProgressCell"];
        cell.indicatorIsHidden = tableView.refreshControl.isRefreshing || [_activeDownloader isDownloaded];
        return cell;
    }
    FeedPostCell * cell = [tableView dequeueReusableCellWithIdentifier:@"PostCell"];
    cell.delegate = self;
    cell.post = (EGFPost *)[_activeDownloader objectAtIndex:indexPath.row];
    return cell;
}
@end

File: GuideSwift/GuideSwift/ViewControllers/FeedController/FeedController.swift
//
//  FeedController.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 12.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit
import EGF2

class FeedController: BaseTableController, PostCellDelegate, UISearchBarDelegate, BaseDownloaderDelegate {

    @IBOutlet var searchButton: UIBarButtonItem!
    @IBOutlet var newPostButton: UIBarButtonItem!
    @IBOutlet var listSegment: UISegmentedControl!
    
    fileprivate var searching: SearchDownloader<EGFPost>?
    fileprivate var timeline: EdgeDownloader<EGFPost>?
    fileprivate var feed: EdgeDownloader<EGFPost>?
    fileprivate let searchBar = UISearchBar()
    fileprivate var cellHeights = [String: CGFloat]()
    fileprivate var currentUserId: String?
    fileprivate let expand = ["creator","image"]
    fileprivate let edge = "posts"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.register(UINib(nibName: "ProgressCell", bundle: nil), forCellReuseIdentifier: "ProgressCell")
        tableView.register(UINib(nibName: "PostCell", bundle: nil), forCellReuseIdentifier: "PostCell")
        
        let parameters = EGF2SearchParameters(withObject: "post")
        parameters.fields = ["desc"]
        parameters.expand = expand
        searching = SearchDownloader(withParameters: parameters)
        searching?.delegate = self
        
        searchBar.delegate = self
        searchBar.showsCancelButton = true
        searchBar.tintColor = UIColor.hexColor(0x5E66B1)
        
        Graph.userObject { (object, error) in
            guard let user = object as? EGFUser, let userId = user.id else { return }
            self.currentUserId = userId
            self.timeline = EdgeDownloader(withSource: userId, edge: "timeline", expand: self.expand)
            self.timeline?.delegate = self
            self.feed = EdgeDownloader(withSource: userId, edge: self.edge, expand: self.expand)
            self.feed?.delegate = self
            self.feed?.tableView = self.tableView
            self.feed?.getNextPage()
            self.activeDownloader = self.feed
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        if !(searchBar.text ?? "").isEmpty {
            searchBar.becomeFirstResponder()
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let postController = segue.destination as? PostController, let indexPath = sender as? IndexPath {
            postController.currentPost = activeDownloader?[indexPath.row]
            searchBar.resignFirstResponder()
        }
    }
    
    fileprivate var activeDownloader: BaseDownloader<EGFPost>? {
        willSet {
            if let oldDownloader = activeDownloader {
                oldDownloader.tableView = nil
            }
        }
        didSet {
            activeDownloader?.tableView = tableView
        }
    }
    
    @IBAction func changeList(_ sender: UISegmentedControl) {
        if sender.selectedSegmentIndex == 0 {
            activeDownloader = feed
        }
        else if sender.selectedSegmentIndex == 1 {
            activeDownloader = timeline
        }
    }
    
    // MARK:- BaseDownloaderDelegate
    func willUpdate(graphObject: NSObject) {
        guard let objectId = graphObject.value(forKey: "id") as? String else { return }
        self.cellHeights.removeValue(forKey: objectId)
    }
    
    // MARK:- Searching
    @IBAction func beginSearch(_ sender: AnyObject) {
        activeDownloader = searching
        navigationItem.rightBarButtonItem = nil
        navigationItem.leftBarButtonItem = nil
        navigationItem.titleView = searchBar
        searchBar.becomeFirstResponder()
    }
    
    func endSearch() {
        changeList(listSegment)
        searchBar.text = nil
        navigationItem.rightBarButtonItem = newPostButton
        navigationItem.leftBarButtonItem = searchButton
        navigationItem.titleView = listSegment
    }
    
    // MARK:- UISearchBarDelegate
    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        endSearch()
    }
    
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        if let text = searchBar.text, text.isEmpty {
            endSearch()
        }
    }
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        searching?.showObjects(withQuery: searchText)
    }
    
    // MARK:- PostCellDelegate
    var authorizedUserId: String? {
        get {
            return currentUserId
        }
    }
    
    func delete(post: EGFPost) {
        guard let postId = post.id, let userId = self.currentUserId else { return }
        
        showConfirm(withTitle: "Warning", message: "Really delete?") { 
            ProgressController.show()
            Graph.deleteObject(withId: postId, forSource: userId, fromEdge: self.edge) { (_, error) in
                ProgressController.hide()
            }
        }
    }
    
    // MARK:- UITableViewDelegate
    override func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        if let control = refreshControl, control.isRefreshing == true {
            activeDownloader?.refreshList()
        }
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.section == 0 {
            if let post = activeDownloader?[indexPath.row], let postId = post.id {
                // Check if we already have the value
                if let value = cellHeights[postId] { return value }
                
                let height = PostCell.height(forPost: post)
                cellHeights[postId] = height
                return height
            }
        }
        return 44
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        guard let downloader = activeDownloader else { return }
        if indexPath.section == 1 && !downloader.isDownloaded {
            downloader.getNextPage()
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        performSegue(withIdentifier: "ShowPost", sender: indexPath)
    }
    
    // MARK:- UITableViewDataSource
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return section == 0 ? (activeDownloader?.count ?? 0) : 1
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 1 {
            let cell = tableView.dequeueReusableCell(withIdentifier: "ProgressCell") as! ProgressCell
            cell.indicatorIsHidden = tableView.refreshControl!.isRefreshing || (activeDownloader?.isDownloaded ?? true)
            return cell
        }
        let cell = tableView.dequeueReusableCell(withIdentifier: "PostCell") as! PostCell
        cell.delegate = self
        cell.post = activeDownloader![indexPath.row]
        return cell
    }
}

